### PR TITLE
DBLayer: add bracketObserveIO around SQLite queries

### PR DESCRIPTION
**Pull request by rvl**
Relates to issue #354

# Overview

- Adds timing of database queries ([doc pdf here](https://github.com/input-output-hk/iohk-monitoring-framework/blob/master/docs/IOHK-Monitoring.pdf)).
- Adds a test case with a monitor that checks whether the observed value is greater than 0 (the monitor should perform a logging action every time a query has been run).

# Comments

The `bracketObserveIO` function requires a logging `Configuration` in addition to the tracer object. So there are a few changes following from that.

I can't get the test case to work with [aggregation](https://input-output-hk.github.io/iohk-monitoring-framework/one-pager/pdf/benchmarking.pdf) and [monitoring](https://input-output-hk.github.io/iohk-monitoring-framework/one-pager/pdf/monitoring.pdf).

### CODE REVIEW

**Review by paweljakubas — COMMENTED** [review]
@rvl I wonder if this PR is still to be merged or to be closed in the light of #492 . if the former, then let's rebase it

**Comment by rvl** [comment]
@paweljakubas I have rebased just now. This adds `bracketObserveIO` around queries and #492 added structured logging. It is basically impossible to merge either one of the PRs without causing the other one to conflict.

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 1]
```diff
@@ -168,11 +170,54 @@ simpleSpec = do
...
+            unsafeRunExceptT $ createWallet db testPk testCp testMetadata
+            let walletName = T.unpack $ coerce $ name testMetadata
+            msgs <- T.unlines . map snd . logMessages <$> getLogs
+            T.unpack msgs `shouldNotContain` walletName
+
+    describe "Sqlite observables" $ do
+        it "should measure query timings" $ \(db, getLogs) -> do
+            replicateM_ 5 $ listWallets db
+            msgs <- T.unlines . map snd . logMessages <$> getLogs
+            T.unpack msgs `shouldContain` "runQuery monitor works"
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs]: I see failure of this in CI : 
```
    Failures:
    
      test/unit/Cardano/Wallet/DB/SqliteSpec.hs:191:13: 
      1) Cardano.Wallet.DB.Sqlite, Sqlite observables, should measure query timings
           "SELECT \"wallet_id\" FROM \"wallet\" ORDER BY \"wallet_id\"\nSELECT \"wallet_id\" FROM \"wallet\" ORDER BY \"wallet_id\"\nSELECT \"wallet_id\" FROM \"wallet\" ORDER BY \"wallet_id\"\nSELECT \"wallet_id\" FROM \"wallet\" ORDER BY \"wallet_id\"\nSELECT \"wallet_id\" FROM \"wallet\" ORDER BY \"wallet_id\"\n" does not contain "runQuery monitor works"
```
looks like `newMemoryDBLayer` has not taken **[CreateMessage Info "runQuery monitor works"]** into account

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 2]
```diff
@@ -60,6 +64,8 @@ import Cardano.Wallet.Primitive.Types
     )
 import Cardano.Wallet.Unsafe
     ( unsafeRunExceptT )
+import Control.Monad
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:69]: you should put `replicateM_` with `unless` under one hood here

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 3]
```diff
@@ -168,11 +170,54 @@ simpleSpec = do
...
+        (Just [CM.KatipBK])
+    CM.setMonitors logConfig $ HM.singleton
+        "query.runQuery"
+        ( Just (Compare "ObserveDiff" (GE, (OpMeasurable 0)))
+        , Compare "ObserveDiff" (GE, (OpMeasurable 0))
+        , [CreateMessage Info "runQuery monitor works"]
+        )
+
+    CM.setBackends logConfig
+        "query.runQuery"
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs]: I wonder if setting backends somehow overlaps with setting monitors and that is why we do not see `runQuery monitor works`

Comment (paweljakubas) [reply lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs]: Or maybe as we have `Debug` severity and setting monitors with `CreateMessage Info` is having wrong log severity ....

**Review by paweljakubas — CHANGES_REQUESTED** [review]
After rebase I have finally understood what is going on. There is unit test failing that needs addressing and small import fix to be adopted and we should be fine!

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs (hunk 2 above)
Comment (rvl) [reply lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:69]: OK fixed.

**Comment by rvl** [comment]
I was not able to use a monitor in the test case, so had to settle for checking for the subtrace. It will pass the tests now.

I have raised an issue input-output-hk/iohk-monitoring-framework#391 for the monitoring problem.

**Review by paweljakubas — APPROVED** [review]
Cool! I bet we will wait until monitoring team sort out the problem and then conclude this in the next PR. But until then, let's merge it

**Comment by CodiePP** [comment]
the PR with the monitoring changes has just been merged. Please, have a look at: https://github.com/input-output-hk/iohk-monitoring-framework/pull/399

**Pull request merged**